File: Hanaihang/scripts/scrape-siamparagon-directory.py
import json
import os
import re
from urllib.request import Request, urlopen

DIRECTORY_URL = "https://www.siamparagon.co.th/directory"

def fetch_html():
    req = Request(DIRECTORY_URL, headers={"User-Agent": "Mozilla/5.0"})
    return urlopen(req, timeout=30).read().decode("utf-8", "ignore")

def extract_payload(html: str) -> str:
    pattern = re.compile(r'self.__next_f.push\(\[(\d+),"(.*?)"\]\)', re.S)
    payload = ""
    for _, chunk in pattern.findall(html):
        payload += chunk.encode("utf-8").decode("unicode_escape")
    return payload

def extract_occupants(payload: str) -> dict:
    key = '"occupantsByCategory":'
    idx = payload.find(key)
    if idx == -1:
        raise RuntimeError("occupantsByCategory not found in payload")
    start = idx + len(key)
    while start < len(payload) and payload[start].isspace():
        start += 1
    if payload[start] != '{':
        raise RuntimeError("Expected object for occupantsByCategory")
    level = 0
    end = start
    for i in range(start, len(payload)):
        ch = payload[i]
        if ch == '{':
            level += 1
        elif ch == '}':
            level -= 1
            if level == 0:
                end = i + 1
                break
    return json.loads(payload[start:end])

def normalize_level(level: str) -> str:
    if not level:
        return ""
    cleaned = str(level).strip().upper()
    if cleaned == "GF":
        return "G"
    return cleaned

def main():
    html = fetch_html()
    payload = extract_payload(html)
    occupants = extract_occupants(payload)

    list_items = []
    for category in ["shop", "dine", "seeAndDo"]:
        for entry in occupants.get(category, []):
            props = entry.get("properties", {})
            name = (props.get("name", {}) or {}).get("en", "").strip()
            floor = normalize_level(props.get("level_name", ""))
            keywords = props.get("keywords", [])
            list_items.append({
                "name": name,
                "floor": floor,
                "category": category,
                "keywords": keywords,
            })

    output = {
        "source": DIRECTORY_URL,
        "retrievedAt": __import__("datetime").datetime.utcnow().isoformat() + "Z",
        "count": len(list_items),
        "stores": [item for item in list_items if item["name"]],
    }

    out_dir = os.path.join(os.getcwd(), "data", "derived")
    os.makedirs(out_dir, exist_ok=True)
    out_path = os.path.join(out_dir, "siamparagon-directory.json")
    with open(out_path, "w", encoding="utf-8") as f:
        json.dump(output, f, ensure_ascii=False, indent=2)

    print(f"Extracted {output['count']} entries")
    print(f"Output: {out_path}")

if __name__ == "__main__":
    main()

File: Hanaihang/scripts/scrape-siamparagon-directory.mjs
import fs from 'node:fs';
import path from 'node:path';
import process from 'node:process';

const DIRECTORY_URL = 'https://www.siamparagon.co.th/directory';

function decodeEscaped(input) {
  return input.replace(/\\u([0-9a-fA-F]{4})/g, (_, code) =>
    String.fromCharCode(parseInt(code, 16))
  );
}

async function fetchHtml() {
  const res = await fetch(DIRECTORY_URL, {
    headers: {
      'user-agent': 'Mozilla/5.0 (compatible; HanaihangBot/1.0)',
    },
  });
  if (!res.ok) {
    throw new Error(`Failed to fetch directory: ${res.status}`);
  }
  return res.text();
}

function extractPayload(html) {
  const regex = /self.__next_f.push\(\[(\d+),"(.*?)"\]\)/gs;
  const chunks = [];
  let match;
  while ((match = regex.exec(html)) !== null) {
    chunks.push(decodeEscaped(match[2]));
  }
  return chunks.join('');
}

function extractOccupants(payload) {
  const key = 'occupantsByCategory';
  const idx = payload.indexOf(key);
  if (idx === -1) {
    throw new Error('occupantsByCategory not found in payload');
  }

  // find the first '{' after the key
  let start = payload.indexOf('{', idx);
  if (start === -1) {
    throw new Error('Expected object for occupantsByCategory');
  }

  let level = 0;
  let end = start;
  for (let i = start; i < payload.length; i += 1) {
    const ch = payload[i];
    if (ch === '{') level += 1;
    if (ch === '}') {
      level -= 1;
      if (level === 0) {
        end = i + 1;
        break;
      }
    }
  }

  const jsonStr = payload.slice(start, end);
  const data = JSON.parse(jsonStr);
  return data;
}

function normalizeLevel(level) {
  if (!level) return '';
  const cleaned = String(level).trim().toUpperCase();
  if (cleaned === 'GF') return 'G';
  return cleaned;
}

async function main() {
  const html = await fetchHtml();
  const payload = extractPayload(html);
  const occupants = extractOccupants(payload);

  const list = [];
  const categories = ['shop', 'dine', 'seeAndDo'];
  for (const category of categories) {
    const entries = occupants[category] || [];
    for (const entry of entries) {
      const props = entry.properties || {};
      const name = props.name?.en?.trim() || '';
      const floor = normalizeLevel(props.level_name);
      const keywords = props.keywords || [];
      list.push({
        name,
        floor,
        category,
        keywords,
      });
    }
  }

  const output = {
    source: DIRECTORY_URL,
    retrievedAt: new Date().toISOString(),
    count: list.length,
    stores: list.filter((s) => s.name),
  };

  const outDir = path.join(process.cwd(), 'data', 'derived');
  fs.mkdirSync(outDir, { recursive: true });
  const outPath = path.join(outDir, 'siamparagon-directory.json');
  fs.writeFileSync(outPath, JSON.stringify(output, null, 2));

  console.log(`Extracted ${output.count} entries`);
  console.log(`Output: ${outPath}`);
}

main().catch((error) => {
  console.error('Scrape failed:', error);
  process.exit(1);
});
